### PR TITLE
ARROW-12246: [CI] Sync conda recipes with upstream feedstock

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
(I am first trying with linux)

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
@github-actions crossbow submit -g conda

**Comment by jorisvandenbossche** [comment]
@github-actions crossbow submit conda-linux-gcc-py38-cpu

**Comment by xhochy** [comment]
@github-actions crossbow submit conda-linux-gcc-py38-cpu

**Comment by jorisvandenbossche** [comment]
I *suppose* that in the past we made some adjustments to the build scripts (`run_docker_build.sh`) after copying them from upstream conda-forge, to make it work here? (cc @kszucs) 
If that's the case, it would be good to better document that / more clearly separate those changes (so they can be applied later again).

**Comment by jorisvandenbossche** [comment]
Eg https://github.com/apache/arrow/pull/4649 has some changes to that file after the initial commit adding it: https://github.com/apache/arrow/pull/4649/files/2037f78fc8a5842a123930656afe73f7a1e25cbb..20f6cecd3e8a8e476387a6441d6572ff94fa6e5f#diff-7ad035b7e05f08778571fe6908954a873b6ca24fe661324cb9d1a29c7c4e505b

**Comment by xhochy** [comment]
@jorisvandenbossche I would revert the changes to `dev/tasks/conda-recipes/{run_docker_build.sh,build_steps.sh}` and put a comment at the top that we did modify them manually. We shouldn't need to update them. Also I would have a bit of time today to finish this PR, ok if I take this over from you?

**Comment by jorisvandenbossche** [comment]
> ok if I take this over from you?

Sure!

**Comment by xhochy** [comment]
@github-actions crossbow submit conda-linux-gcc-py38-cpu

**Comment by kszucs** [comment]
Note that I have a large undertaking for crossbow #9913 which I plan to merge today, so expect for some minor conflicts. 

@xhochy Just an idea: could we `clone` the upstream conda-forge repository in the crossbow task instead of vendoring the files? We would need to vendor a couple of fines to enable us introducing modifying dependencies, but this approach could help to keep the setup in sync with the feedstock.

**Comment by xhochy** [comment]
> Note that I have a large undertaking for crossbow #9913 which I plan to merge today, so expect for some minor conflicts.

I don't see any conflicting things here, so I would hope that a rebase here will work out fine. Thanks for the heads-up 👍 

> @xhochy Just an idea: could we `clone` the upstream conda-forge repository in the crossbow task instead of vendoring the files? We would need to vendor a couple of fines to enable us introducing modifying dependencies, but this approach could help to keep the setup in sync with the feedstock.

Yes, that could be part of a future solution. I'm currently struggling a bit with it though as the number of possible conda configurations varies on conda-forge and the names of the files in `.ci_support` changes quite often. Furthermore, it would be nice to have a `libarrow` recipe that is the base for both the Python and R packages but doesn't depend on Python. This would need a bit more of a restructuring inside the arrow codebase.

**Comment by xhochy** [comment]
@github-actions crossbow submit -g conda

**Comment by xhochy** [comment]
@github-actions crossbow submit conda-win-*

**Comment by xhochy** [comment]
@jorisvandenbossche @kszucs This is ready for review, updated all x86 conda packages and sneaked in builds for the Apple M1.

**Comment by xhochy** [comment]
One of the Windows builds progressed so far on my crossbow that I'm confident that the rest will also be green in ~1h: https://dev.azure.com/mail0406/crossbow/_build/results?buildId=26&view=logs&j=4c86bc1b-1091-5192-4404-c74dfaad23e7&t=41795ef0-6501-5db4-3ad4-33c0cf085626

**Comment by xhochy** [comment]
@github-actions crossbow submit -g conda

File: dev/tasks/conda-recipes/arrow-cpp/bld-pyarrow.bat [hunk 1]
```diff
@@ -1,16 +1,17 @@
...
-del cmake_modules\SetupCxxFlags.cmake
-del cmake_modules\CompilerInfo.cmake
-del cmake_modules\FindNumPy.cmake
-del cmake_modules\FindPythonLibsNew.cmake
-copy /Y "%SRC_DIR%\cpp\cmake_modules\BuildUtils.cmake" cmake_modules\
-copy /Y "%SRC_DIR%\cpp\cmake_modules\SetupCxxFlags.cmake" cmake_modules\
-copy /Y "%SRC_DIR%\cpp\cmake_modules\CompilerInfo.cmake" cmake_modules\
-copy /Y "%SRC_DIR%\cpp\cmake_modules\FindNumPy.cmake" cmake_modules\
-copy /Y "%SRC_DIR%\cpp\cmake_modules\FindPythonLibsNew.cmake" cmake_modules\
+@rem del cmake_modules\BuildUtils.cmake
```
Comment (jorisvandenbossche) [inline dev/tasks/conda-recipes/arrow-cpp/bld-pyarrow.bat:5]: Worth adding a comment that those are commented out compared to the upstream conda-forge build script?

File: dev/tasks/conda-recipes/arrow-cpp/meta.yaml [hunk 2]
```diff
@@ -1,6 +1,6 @@
-{% set version = "3.0.0" %}
+{% set version = ARROW_VERSION %}
```
Comment (jorisvandenbossche) [inline dev/tasks/conda-recipes/arrow-cpp/meta.yaml:1]: Maybe can add a comment here indicating that setting this to `ARROW_VERSION` is a difference compared to the upstream version?

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks for finishing this! Assuming the builds pass, this is all good for me. 
I just added two suggestions of adding some more comments for our future selves (I commented on a subset of the diff, to see the changes after updating with the upstream version, but could be I was looking at wrong changes;)).

**Pull request closed without merging**